src/datahike/index/hitchhiker_tree/upsert.cljc

```diff
@@ -2,54 +2,86 @@
   (:require [hitchhiker.tree :as tree]
             [hitchhiker.tree.op :as op]))
 
+(defn- increase-by-one?
+  "Returns true if elements in vector 'indices' form a prefix of the vector indices.
+  This is equivalent to check whether the elements increase by one and contains 0.
+  E.g., [0 1 2] => true, [0 2 3] => false, [1 2] => false."
+  [indices]
+  (if (.contains indices 0)
+    (let [m (apply max indices)
+          s (apply + indices)]
+      (= s (/ (* m (+ 1 m)) 2)))
+    false))
+
+(def prefix? (memoize increase-by-one?))
+
+(defn- max-t
+  "Returns the key with max 't' component."
+  [ks]
+  (when (seq ks)
+    (apply max-key #(nth % 3) ks)))
+
+(defn mask [new indices]
+  (reduce (fn [mask pos]
+            (assoc mask pos (nth new pos)))
+          [nil nil nil nil]
+          indices))
+
 (defn old-key
   "Returns the old version of the given 'new' key if it exists in 'old-keys'.
-  If there are multiple old versions, the one with the biggest transaction time is returned."
-  [old-keys new]
-  (let [[a b _ _] new]
-    (when (seq old-keys)
-      (when-let [candidates (subseq old-keys >= [a b nil nil])]
+  If there are multiple old versions, the one with the biggest transaction time is returned.
+  'indices' is a vector of integer indicating which positions in keys are significant,
+  i.e., [0 2] means that the first and third entry in the key are used for filtering."
+  [old-keys new indices]
+  (when (seq old-keys)
+    (let [mask (mask new indices)]
+      (when-let [candidates (subseq old-keys >= mask)]
         (->> candidates
              (map first)
-             (take-while #(and (= a (first %)) (= b (second %))))
-             reverse
-             first)))))
+             ((if (prefix? indices) take-while filter)
```

Comment:
What is going on here?
When the 'indices' in the vector are strictly increasing by 1, like in [0 1 2], then we know that we are dealing with results returned from 'subseq' which are sorted from the 1st, 2nd and third entries in the datom. Thus we can use 'take-while' to retrieve all matching datoms until there is no match.
But when the 'indices' vector is like [0 2], which is the case for :avet, 'subseq' won't return datoms sorted with 'a' then 'e' but rather with 'a' then 'v' and 'e'. Thus we can't use 'take-while' but are forced to filter out all the results from 'subseq' that match 'a' and 'e'.
In summary, 'prefix?' is used to check whether the 'indices' vector represents an index where the prefix of datoms are used to sort them out or not.

Reply:
Ok, thank you for explaining!

Reply:
Ok. Thank you for explaining!